spmd/tensor/api.py

```diff
@@ -143,29 +161,29 @@ def __new__(
             already have tensor initialized and want to shard this tensor),
             consider using `distribute_tensor`.
         """
-        # recover tensor shape and strides in the case of sharding
-        tensor_shape = list(local_tensor.size())
+        # recover tensor strides from local tensor strides and global size info
+        # in the case of sharding
         tensor_stride = list(local_tensor.stride())
-        for idx, placement in enumerate(placements):
+        local_size = list(local_tensor.size())
+        for placement in placements:
             if isinstance(placement, Shard):
                 shard_dim = placement.dim
-                local_dim_size = tensor_shape[shard_dim]
-                # recover tensor shape on the shard dim
-                tensor_shape[shard_dim] = local_dim_size * device_mesh.size(idx)
-
                 # recover tensor stride by modifying the stride that larger than
                 # the current stride on the shard_dim
                 for i in range(len(tensor_stride)):
-                    if tensor_stride[i] > tensor_stride[shard_dim]:
-                        tensor_stride[i] = tensor_stride[i] * device_mesh.size(
-                            idx
-                        )
+                    if (
+                        i != shard_dim
+                        and tensor_stride[i] >= tensor_stride[shard_dim]
+                    ):
+                        # rescale the stride by the shard size
+                        tensor_stride[i] = (
+                            tensor_stride[i] // local_size[shard_dim]
+                        ) * size[shard_dim]
```

Comment:
How does this actually work? Are we creating a tensor with custom strides? In case the tensor is not contiguous, what happens in practice?

Reply:
For uneven sharding, would we need to keep uneven strides?